ducktape/tests/runner_client.py

```diff
@@ -0,0 +1,281 @@
+                    self.log(logging.WARN, "Error cleaning services: %s" % e.message + "\n" + traceback.format_exc(limit=16))
+
+        try:
+            self.test.free_nodes()
+        except BaseException as e:
+            exceptions.append(e)
+            self.log(logging.WARN, "Error freeing nodes: %s" % e.message + "\n" + traceback.format_exc(limit=16))
+
+        # Remove reference to services. This is important to prevent potential memory leaks if users write services
+        # which themselves have references to large memory-intensive objects
+        del self.test_context.services
+
+    def log(self, log_level, msg):
+        """Log to the service log and the test log of the current test."""
+
+        if not getattr(self, 'test'):
+            msg = "%s: %s" % (self.__class__.__name__, str(msg))
+            self.logger.log(log_level, msg)
+        else:
+            msg = "%s: %s: %s" % (self.__class__.__name__, self.test_context.test_name, str(msg))
+            self.logger.log(log_level, msg)
+            self.test.logger.log(log_level, msg)
+
+        self.send(self.request.log(msg, level=log_level))
+
+
+class Sender(object):
+    REQUEST_TIMEOUT_MS = 3000
+    NUM_RETRIES = 5
+
+    def __init__(self, server_host, server_port, logger):
+        self.serde = SerDe()
+        self.server_endpoint = "tcp://%s:%s" % (str(server_host), str(server_port))
+        self.zmq_context = zmq.Context()
+        self.socket = None
+        self.poller = zmq.Poller()
+
+        self.logger = logger
+
+        self._init_socket()
+
+    def _init_socket(self):
+        self.socket = self.zmq_context.socket(zmq.REQ)
+        self.socket.connect(self.server_endpoint)
+        self.poller.register(self.socket, zmq.POLLIN)
+
+    def send(self, message, blocking=True):
+
+        retries_left = Sender.NUM_RETRIES
+        serialized_event = self.serde.serialize(message)
+
+        while retries_left > 0:
+            self.logger.debug("client: sending event: " + str(message))
+            self.socket.send(serialized_event)
+            retries_left -= 1
+            waiting_for_reply = True
+
+            while waiting_for_reply:
+                self.logger.debug("polling for response")
+                sockets = dict(self.poller.poll(Sender.REQUEST_TIMEOUT_MS))
+
+                if sockets.get(self.socket) == zmq.POLLIN:
+                    self.logger.debug("POLLIN")
+                    reply = self.socket.recv()
+                    if reply:
+                        self.logger.debug("received", self.serde.deserialize(reply))
+                        return self.serde.deserialize(reply)
+                    else:
+                        # send another request...
+                        break
+                else:
+                    self.logger.debug("NO-POLLIN")
+                    self.close()
+                    self._init_socket()
+                    waiting_for_reply = False
+                time.sleep(.5)
```

Comment:
Why `time.sleep`? Why not just block indefinitely or set `REQUEST_TIMEOUT_MS` much, much larger? It looks like maybe zmq doesn't do anything in terms of reliability by default. This is pretty concerning... Knowing that failures can occur is good, and having a way to actually handle them gracefully is better, but firing and praying doesn't seem great, especially since I'm not sure all the code on the other half of this connection is idempotent. In fact, the logging code *definitely* isn't idempotent.